### PR TITLE
LUI-127 | Fix porxy error is thrown when user tries to rebuild search…

### PR DESCRIPTION
**Pull request by sumanmaity112**

### CODE REVIEW

File: omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java [hunk 1]
```diff
@@ -48,12 +51,33 @@ public String showPage() {
...
 		} catch (RuntimeException e) {
 			success = false;
 		}
 
 		results.put("success", success);
 		return results;
 	}
+
+	/**
+	 * @should return return inProgress for status if a rebuildSearchIndex is not completed
```
Comment (sumanmaity112) [inline omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java]: This return was already there, I didn't make any changes

File: omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java [hunk 2]
```diff
@@ -48,12 +51,33 @@ public String showPage() {
...
+    @RequestMapping(method = RequestMethod.GET, value = "admin/maintenance/rebuildSearchIndexStatus")
+    public @ResponseBody Map<String, String> getStatus() {
+    	if(updateSearchIndexAsync == null)
+    		throw new APIException("There was a problem rebuilding the search index");
+
+        Map<String, String> results = new HashMap<>();
+        if (updateSearchIndexAsync.isDone()) {
+			results.put("status", updateSearchIndexAsync.isCancelled() ? "error" : "success");
+            updateSearchIndexAsync = null;
+        } else
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java]: This code would be clearer and safe if you used {} braces even for the else clause as advised here https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

**Comment by dkayiwa** [comment]
Should this work when one is not logged in?

File: omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java (hunk 2 above)
Comment (sumanmaity112) [inline omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java]: will make this change and create a new pr

**Comment by sumanmaity112** [comment]
Only admin can rebuild the search index. So, any user is not logged in, he/she should not be able to access this.

**Comment by dkayiwa** [comment]
I logged out from another tab but i was still able to build the search index

File: omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java [hunk 3]
```diff
@@ -62,8 +74,70 @@ public void rebuildSearchIndex_shouldReturnTrueForSuccessIfTheUpdateDoesNotFail(
...
+        Map<String, String> response = controller.getStatus();
+        assertEquals("error", response.get("status"));
+    }
+
+    /**
+     * @verifies throws API exception if getStatus called before rebuildSearchIndex
+     * @see SearchIndexController#getStatus()
+     */
+    @Test()
+    public void getStatus_shouldThrowApiException() throws Exception {
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: Should it throw the exception all the time its called?

File: omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java [hunk 4]
```diff
@@ -62,8 +74,70 @@ public void rebuildSearchIndex_shouldReturnTrueForSuccessIfTheUpdateDoesNotFail(
...
+        expectedException.expect(APIException.class);
+        expectedException.expectMessage("There was a problem rebuilding the search index");
+        controller.getStatus();
+        Assert.fail();
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: What is the use of the above Assert.fail?

File: omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java [hunk 5]
```diff
@@ -62,8 +84,88 @@ public void rebuildSearchIndex_shouldReturnTrueForSuccessIfTheUpdateDoesNotFail(
...
+        Map<String, Object> response = controller.rebuildSearchIndex();
+        assertEquals(false, response.get("success"));
+    }
+
+    /**
+     * @verifies return false for success if a deny user makes call
+     * @see SearchIndexController#rebuildSearchIndex()
+     */
+    @Test
+    public void rebuildSearchIndex_shouldReturnFalseForSuccessIfADenyUserMakesCall() throws Exception {
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: Looking at the method name, what does a deny user mean?

File: omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java [hunk 6]
```diff
@@ -47,13 +50,39 @@ public String showPage() {
 		boolean success = true;
 		Map<String, Object> results = new HashMap<String, Object>();
 		log.debug("rebuilding search index");
-		try {
-			Context.updateSearchIndex();
-		} catch (RuntimeException e) {
+		if (!Context.getUserContext().isAuthenticated()) {
 			success = false;
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java]: So when a non authenticated user tries to build the index, does it just fail, or does it tell them that they have to first login, or does it redirect them to the login page?

Comment (sumanmaity112) [inline omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java]: it just fails as before.

File: omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java (hunk 5 above)
Comment (sumanmaity112) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: Here deny user mean who is not authenticated to the system

Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: Which is easier to understand? Deny user or un authenticated user? 😊

Comment (sumanmaity112) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: **un-authenticated user** obviously. should I change it?

Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: Oh yes!

Comment (sumanmaity112) [inline omod/src/test/java/org/openmrs/web/controller/maintenance/SearchIndexControllerTest.java]: I changed it

**Pull request merged**